File: app/src/test/java/com/concrete/challenge/githubjavapop/ui/repository/PullRequestViewModelTest.java
package com.concrete.challenge.githubjavapop.ui.repository;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.LifecycleRegistry;
import androidx.lifecycle.Observer;

import com.concrete.challenge.githubjavapop.api.Api;
import com.concrete.challenge.githubjavapop.api.SingleSchedulers;
import com.concrete.challenge.githubjavapop.domain.PullRequest;
import com.concrete.challenge.githubjavapop.ui.pull.PullRequestViewModel;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.ArrayList;

import io.reactivex.Single;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class PullRequestViewModelTest {

    @Rule
    public InstantTaskExecutorRule instantExecutorRule = new InstantTaskExecutorRule();

    @Mock
    Api api;
    PullRequestViewModel pullRequestViewModel;
    @Mock
    Observer<ArrayList<PullRequest>> pullRequestObserver;
    @Mock
    Observer<Throwable> errorObserver;
    @Mock
    LifecycleOwner lifecycleOwner;
    Lifecycle lifecycle;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
        lifecycle = new LifecycleRegistry(lifecycleOwner);
        pullRequestViewModel = new PullRequestViewModel(api, SingleSchedulers.TEST);
        pullRequestViewModel.getPullRequests().observeForever(pullRequestObserver);
        pullRequestViewModel.getError().observeForever(errorObserver);
    }

    @Test
    public void testObserverNull() {
        when(api.getPullRequests("", "", 1)).thenReturn(null);
        assertNotNull(pullRequestViewModel.getPullRequests());
        assertTrue(pullRequestViewModel.getPullRequests().hasObservers());
    }

    @Test
    public void testApiLoadSuccess() {
        ArrayList<PullRequest> pullRequests = new ArrayList<>();
        pullRequests.add(new PullRequest());

        when(api.getPullRequests("", "", 1)).thenReturn(Single.just(pullRequests));
        pullRequestViewModel.loadPullRequests("", "", 1);
        verify(pullRequestObserver).onChanged(pullRequests);
    }

    @Test
    public void testApiLoadFailed() {
        Throwable throwable = new Throwable();

        when(api.getPullRequests("", "", 1)).thenReturn(Single.error(throwable));
        pullRequestViewModel.loadPullRequests("", "", 1);
        verify(errorObserver).onChanged(throwable);
    }

    @After
    public void destroy() {
        api = null;
        pullRequestViewModel = null;
    }
}

File: app/src/test/java/com/uharris/desafio_android/presentation/PullRequestViewModelTest.kt
package com.uharris.desafio_android.presentation

import android.app.Application
import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.lifecycle.Observer
import com.nhaarman.mockito_kotlin.*
import com.uharris.desafio_android.data.base.Failure
import com.uharris.desafio_android.data.base.Result
import com.uharris.desafio_android.domain.models.PullRequest
import com.uharris.desafio_android.domain.usecases.actions.FetchPullRequest
import com.uharris.desafio_android.factory.DataFactory
import com.uharris.desafio_android.factory.ResponseFactory
import com.uharris.desafio_android.presentation.sections.pull.PullRequestViewModel
import com.uharris.desafio_android.presentation.state.Resource
import com.uharris.desafio_android.presentation.state.ResourceState
import kotlinx.coroutines.runBlocking
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.mockito.Captor
import org.mockito.Mock
import org.mockito.Mockito

@RunWith(JUnit4::class)
class PullRequestViewModelTest {
    @get:Rule
    var instantTaskExecutorRule = InstantTaskExecutorRule()

    @Mock
    private lateinit var fetchPullRequest: FetchPullRequest
    @Mock
    private lateinit var observerPullRequests: Observer<Resource<List<PullRequest>>>

    @Captor
    private lateinit var captorPullRequest: KArgumentCaptor<(Result<List<PullRequest>>) -> Unit>

    private lateinit var pullRequestViewModel: PullRequestViewModel

    @Before
    fun setup(){
        fetchPullRequest = mock()
        observerPullRequests = mock()

        captorPullRequest = argumentCaptor()

        val application = Mockito.mock(Application::class.java)

        pullRequestViewModel = PullRequestViewModel(fetchPullRequest, application)
    }

    @Test
    fun fetchPullRequestsReturnLoadingState(){
        pullRequestViewModel.pullRequestsLiveData.observeForever(observerPullRequests)

        pullRequestViewModel.fetchPullRequest(DataFactory.randomString(), DataFactory.randomString())

        assert(pullRequestViewModel.pullRequestsLiveData.value?.status == ResourceState.LOADING)
    }

    @Test
    fun fetchPullRequestsReturnSuccessData(){
        val data = ResponseFactory.makeullRequestList()
        val result = Result.Success(data)
        pullRequestViewModel.pullRequestsLiveData.observeForever(observerPullRequests)

        pullRequestViewModel.fetchPullRequest(DataFactory.randomString(), DataFactory.randomString())

        runBlocking { verify(fetchPullRequest).invoke(any(), captorPullRequest.capture())}
        captorPullRequest.firstValue.invoke(result)

        assert(pullRequestViewModel.pullRequestsLiveData.value?.data == data)
    }

    @Test
    fun fetchPullRequestsReturnFailure(){
        pullRequestViewModel.pullRequestsLiveData.observeForever(observerPullRequests)

        pullRequestViewModel.fetchPullRequest(DataFactory.randomString(), DataFactory.randomString())

        runBlocking { verify(fetchPullRequest).invoke(any(), captorPullRequest.capture()) }
        captorPullRequest.firstValue.invoke(Result.Error(Failure.ServerError))

        assert(pullRequestViewModel.pullRequestsLiveData.value?.status == ResourceState.ERROR)
    }
}